File: because/dungeon.py
from geom import Vec2D

class DungeonTerrain:
    def __init__(self, width: int, height: int):
        self._tiles = []*width*height
        self.xmax = width
        self.ymax = height

    def tiles(self):
        return iter(self._tiles)

    def __getitem__(self, pos):
        if isinstance(pos, Vec2D):
            return self._tiles[pos.y*self.xmax + pos.x]
        else:
            return self._tiles[pos[1]*self.xmax + pos[0]]

    def __setitem__(self, pos, tile):
        if isinstance(pos, Vec2D):
            self._tiles[pos.y*self.xmax + pos.x] = tile
        else:
            self._tiles[pos[1]*self.xmax + pos[0]] = tile


class DungeonLevel:
    def __init__(self, terrain, entities):
        self.terrain = terrain
        self.entities = entities

File: shared/src/main/scala/com/electronwill/why/Dungeon.scala
package com.electronwill.why

class DungeonLevel(val level: Int,
                   val name: String,
                   val spawnPosition: Vec2i,
                   val terrain: DungeonTerrain) {
  def entityAt(pos: Vec2i): Option[Entity] = None // TODO implement entities
}

class DungeonTerrain(width: Int, height: Int) {
  private val grid = Grid[Tile](width, height)

  export grid._ // expose grid methods as DungeonTerrain's
}
